cmd/sync-tags/main.go

```diff
@@ -199,14 +214,35 @@ func main() {
 			continue
 		}
 
-		// skip if it already exists in origin
-		if _, found := bTagCommits[bName]; found {
+		// skip if either tag exists at origin
+		_, nonSemverTagAtOrigin := bTagCommits[bName]
+		_, semverTagAtOrigin := bTagCommits[semverTag]
+		if nonSemverTagAtOrigin || (publishSemverTag && semverTagAtOrigin) {
 			continue
 		}
 
-		// do not override tags (we build master first, i.e. the x.y.z-alpha.0 tag on master will not be created for feature branches)
-		if tagExists(r, bName) {
-			continue
+		// if any of the tag exists locally,
+		// delete the tags, clear the cache and recreate them
+		if tagExists(r, bName) || (publishSemverTag && tagExists(r, semverTag)) {
+			if err := deleteTag(bName); err != nil {
+				glog.Fatalf("Failed to delete tag %s: %v", bName, err)
+			}
```

Comment:
won't this fail if the condition that got us into this block was `publishSemverTag && tagExists(r, semverTag)`?

Reply:
I expected something more like:
```
if tagExists(r, bName) {
  if err := deleteTag(bName); err != nil {
     glog.Fatalf("Failed to delete tag %s: %v", bName, err)
  }
  if err := cleanCacheForTag(bName); err != nil {
    glog.Fatalf("Failed to clean go mod cache for %s: %v", bName, err)
  }
}
if publishSemverTag && tagExists(r, semverTag) {
  if err := deleteTag(semverTag); err != nil {
     glog.Fatalf("Failed to delete tag %s: %v", semverTag, err)
  }
  if err := cleanCacheForTag(semverTag); err != nil {
    glog.Fatalf("Failed to clean go mod cache for %s: %v", semverTag, err)
  }
}
```

Reply:
it also looks like we should call cleanCacheForTag first, otherwise we won't enter this block if called again after a failure while calling cleanCacheForTag (the git tag will be gone and we won't enter this block)

Reply:
Addressed these comments. PTAL.

I think there were two more things to consider:

- We can't call `cleanCacheForTag(bName)` because the cache would contain pseudoversion.mod, pseudoversion.zip files, not bName.mod, etc files. We'd need to use `cleanCacheForTag(pseudoversion)`.

- We can't call `cleanCacheForTag(pseudoversion)` if bName exists -- we'd also need to check if we are publishing semver tags. In case of semver tagging, we'll have semverTag.mod files in the cache, instead of pseudoversion.mod files.

Reply:
> We can't call cleanCacheForTag(bName) because the cache would contain pseudoversion.mod, pseudoversion.zip files

Good point

> We can't call cleanCacheForTag(pseudoversion) just if bName exists -- we'd also need to check if we are publishing semver tags

cleanCacheForTag (correctly) tolerates absence of any of the files it cares about, so we should be able to call `cleanCacheForTag(pseudoversion)` unconditionally, right?

Reply:
> cleanCacheForTag (correctly) tolerates absence of any of the files it cares about, so we should be able to call cleanCacheForTag(pseudoversion) unconditionally, right?

done